pageserver/src/tenant/mgr.rs

```diff
@@ -1101,101 +1305,436 @@ pub(crate) async fn attach_tenant(
-/// entry is vacant. The closure is responsible for creating the tenant object and inserting
-/// it into the tenants map through the vacnt entry that it receives as argument.
+/// Guards a particular tenant_id's content in the TenantsMap.  While this
+/// structure exists, the TenantsMap will contain a [`TenantSlot::InProgress`]
+/// for this tenant, which acts as a marker for any operations targeting
+/// this tenant to retry later, or wait for the InProgress state to end.
+///
+/// This structure enforces the important invariant that we do not have overlapping
+/// tasks that will try use local storage for a the same tenant ID: we enforce that
+/// the previous contents of a slot have been shut down before the slot can be
+/// left empty or used for something else
 ///
-/// NB: the closure should return quickly because the current implementation of tenants map
-/// serializes access through an `RwLock`.
-async fn tenant_map_insert<F, R>(
+/// Holders of a SlotGuard should explicitly dispose of it, using either `upsert`
+/// to provide a new value, or `revert` to put the slot back into its initial
+/// state.  If the SlotGuard is dropped without calling either of these, then
+/// we will leave the slot empty if our `old_value` is already shut down, else
+/// we will replace the slot with `old_value` (equivalent to doing a revert).
+///
+/// The `old_value` may be dropped before the SlotGuard is dropped, by calling
+/// `drop_old_value`.  It is an error to call this without shutting down
+/// the conents of `old_value`.
+pub struct SlotGuard {
     tenant_id: TenantId,
-    insert_fn: F,
-) -> Result<Arc<Tenant>, TenantMapInsertError>
-where
-    F: FnOnce() -> R,
-    R: std::future::Future<Output = anyhow::Result<Arc<Tenant>>>,
-{
-    let mut guard = TENANTS.write().await;
-    let m = match &mut *guard {
-        TenantsMap::Initializing => return Err(TenantMapInsertError::StillInitializing),
-        TenantsMap::ShuttingDown(_) => return Err(TenantMapInsertError::ShuttingDown),
-        TenantsMap::Open(m) => m,
-    };
-    match m.entry(tenant_id) {
-        hash_map::Entry::Occupied(e) => match e.get() {
-            TenantSlot::Attached(t) => Err(TenantMapInsertError::TenantAlreadyExists(
-                tenant_id,
-                t.current_state(),
-            )),
-            TenantSlot::Secondary => Err(TenantMapInsertError::TenantExistsSecondary(tenant_id)),
-        },
-        hash_map::Entry::Vacant(v) => match insert_fn().await {
-            Ok(tenant) => {
-                v.insert(TenantSlot::Attached(tenant.clone()));
-                Ok(tenant)
+    old_value: Option<TenantSlot>,
+    upserted: bool,
+
+    /// [`TenantSlot::InProgress`] carries the corresponding Barrier: it will
+    /// release any waiters as soon as this SlotGuard is dropped.
+    _completion: utils::completion::Completion,
+}
+
+unsafe impl Send for SlotGuard {}
+unsafe impl Sync for SlotGuard {}
+
```

Comment:
No one questioned this?

Reply:
thanks for pointing this out, cross linking #5802 which addresses this.